### PR TITLE
MJAVADOC-731 update parent, get rid of legacy

### PR DESCRIPTION
**Pull request by kwin**

### CODE REVIEW

**Comment by kwin** [comment]
@michael-o Any idea why the version is managed to 3.3.1 in https://github.com/apache/maven-javadoc-plugin/blob/3032299ec99ed5c968fc3c04cd6f1b0eaeefea20/pom.xml#L415? Couldn't we always use the current version for the report as well?

**Comment by michael-o** [comment]
> @michael-o Any idea why the version is managed to 3.3.1 in
> 
> https://github.com/apache/maven-javadoc-plugin/blob/3032299ec99ed5c968fc3c04cd6f1b0eaeefea20/pom.xml#L415
> ? Couldn't we always use the current version for the report as well?

Yes, it should actually work. At least other reporting plugins reuse themselves in the reactor for reporting. One of the problem is that if you haven't installed the plugin at least once locally, site phase won't include the plugin classes in the reactor. Try and should see it fail.

File: src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java [hunk 1]
```diff
@@ -19,17 +19,54 @@
  * under the License.
  */
 
+
```
Comment (kwin) [inline src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java:22]: not sure why it was previously not following the recommended import order: https://maven.apache.org/developers/conventions/code.html#java-code-convention-import-layouts

File: src/main/java/org/apache/maven/plugins/javadoc/resolver/ResourceResolver.java [hunk 2]
```diff
@@ -19,26 +19,29 @@
  * under the License.
  */
 
+
```
Comment (michael-o) [inline src/main/java/org/apache/maven/plugins/javadoc/resolver/ResourceResolver.java:22]: Seems like one empty line too much...

File: src/main/java/org/apache/maven/plugins/javadoc/resolver/ResourceResolver.java [hunk 3]
```diff
@@ -415,14 +412,15 @@ private Collection<Path> resolveAndUnpack( final List<Artifact> artifacts,
...
-                        + a.getType(), a, e );
+                        + a.getExtension(), RepositoryUtils.toArtifact( a ), e );
                 }
             }
             catch ( final ArchiverException e )
             {
                 if ( propagateErrors )
                 {
-                    throw new ArtifactResolutionException( "Failed to unpack: " + a.getId(), a, e );
+                    throw new ArtifactResolutionException( "Failed to unpack: " + a.toString(),
```
Comment (michael-o) [inline src/main/java/org/apache/maven/plugins/javadoc/resolver/ResourceResolver.java:422]: `toString()` will be automatically invoked.

**Review by michael-o — CHANGES_REQUESTED** [review]
Massive..

File: src/main/java/org/apache/maven/plugins/javadoc/resolver/ResourceResolver.java (hunk 3 above)
Comment (kwin) [reply src/main/java/org/apache/maven/plugins/javadoc/resolver/ResourceResolver.java]: Fixed in https://github.com/apache/maven-javadoc-plugin/pull/164/commits/b88e2980334497fbafa0d26dd48fcf136f819dd0.

File: src/main/java/org/apache/maven/plugins/javadoc/resolver/ResourceResolver.java (hunk 2 above)
Comment (kwin) [reply src/main/java/org/apache/maven/plugins/javadoc/resolver/ResourceResolver.java]: Fixed in https://github.com/apache/maven-javadoc-plugin/pull/164/commits/b88e2980334497fbafa0d26dd48fcf136f819dd0.

**Comment by michael-o** [comment]
going through...

**Review by michael-o — CHANGES_REQUESTED** [review]
Does not work for me. Both behind Nexus and direct connect to Central I always get:
```
84170 [ERROR] Tests run: 23, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 49.635 s <<< FAILURE! - in org.apache.maven.plugins.javadoc.JavadocReportTest
84171 [ERROR] org.apache.maven.plugins.javadoc.JavadocReportTest.testTagletArtifacts  Time elapsed: 1.746 s  <<< ERROR!
84172 org.apache.maven.plugin.MojoExecutionException:
84173 An error has occurred in Javadoc report generation: Unable to resolve artifact:groupId = 'org.codehaus.plexus'
84174 artifactId = 'plexus-javadoc'
84175 version = '1.0'
84176 classifier = 'null'
84177     at org.apache.maven.plugins.javadoc.AbstractJavadocMojo.failOnError(AbstractJavadocMojo.java:6994)
84178     at org.apache.maven.plugins.javadoc.JavadocReport.doExecute(JavadocReport.java:351)
84179     at org.apache.maven.plugins.javadoc.AbstractJavadocMojo.execute(AbstractJavadocMojo.java:1860)
84180     at org.apache.maven.plugins.javadoc.JavadocReportTest.testTagletArtifacts(JavadocReportTest.java:1212)
84181     at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
84182     at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
84183     at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
84184     at java.lang.reflect.Method.invoke(Method.java:498)
84185     at junit.framework.TestCase.runTest(TestCase.java:177)
84186     at junit.framework.TestCase.runBare(TestCase.java:142)
84187     at junit.framework.TestResult$1.protect(TestResult.java:122)
84188     at junit.framework.TestResult.runProtected(TestResult.java:142)
84189     at junit.framework.TestResult.run(TestResult.java:125)
84190     at junit.framework.TestCase.run(TestCase.java:130)
84191     at junit.framework.TestSuite.runTest(TestSuite.java:241)
84192     at junit.framework.TestSuite.run(TestSuite.java:236)
84193     at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:90)
84194     at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:377)
...
84202 Caused by: org.apache.maven.reporting.MavenReportException: Unable to resolve artifact:groupId = 'org.codehaus.plexus'
84203 artifactId = 'plexus-javadoc'
84204 version = '1.0'
84205 classifier = 'null'
84206     at org.apache.maven.plugins.javadoc.AbstractJavadocMojo.addTagletsFromTagletArtifacts(AbstractJavadocMojo.java:5721)
84207     at org.apache.maven.plugins.javadoc.AbstractJavadocMojo.addStandardDocletOptions(AbstractJavadocMojo.java:5537)
84208     at org.apache.maven.plugins.javadoc.AbstractJavadocMojo.executeReport(AbstractJavadocMojo.java:2020)
84209     at org.apache.maven.plugins.javadoc.JavadocReport.generate(JavadocReport.java:149)
84210     at org.apache.maven.plugins.javadoc.JavadocReport.doExecute(JavadocReport.java:346)
84211     ... 23 more
84212 Caused by: org.eclipse.aether.resolution.ArtifactResolutionException: Could not find artifact org.codehaus.plexus:plexus-javadoc:jar:1.0
84213     at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:444)
84214     at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifacts(DefaultArtifactResolver.java:246)
84215     at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifact(DefaultArtifactResolver.java:223)
84216     at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveArtifact(DefaultRepositorySystem.java:294)
84217     at org.apache.maven.plugins.javadoc.AbstractJavadocMojo.createAndResolveArtifact(AbstractJavadocMojo.java:3648)
84218     at org.apache.maven.plugins.javadoc.AbstractJavadocMojo.addTagletsFromTagletArtifacts(AbstractJavadocMojo.java:5717)
84219     ... 27 more
84220 Caused by: org.eclipse.aether.transfer.ArtifactNotFoundException: Could not find artifact org.codehaus.plexus:plexus-javadoc:jar:1.0
84221     at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:434)
84222     ... 32 more
84223
```

**Comment by michael-o** [comment]
@kwin Should I test again?

**Comment by kwin** [comment]
It still fails for me locally and on Jenkins and I don't know why exactly. The mrm-maven-plugin should IMHO deliver the missing artifact but the source in https://github.com/apache/maven-javadoc-plugin/tree/master/src/it/mrm/3rdparty seems incomplete to me. I run into 
```
DEBUG] Resolving artifact org.apache.maven.plugins.javadoc.its:lib:jar:1.0 from [mrm-maven-plugin (http://localhost:51146, default, releases+snapshots)]
[DEBUG] Skipped remote request for org.apache.maven.plugins.javadoc.its:lib:jar:1.0, already updated during this session
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for mjavadoc494 1.0-SNAPSHOT:
[INFO] 
[INFO] mjavadoc494 ........................................ FAILURE [  1.394 s]
[INFO] mjavadoc494-module1 ................................ SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  1.531 s
[INFO] Finished at: 2022-09-22T18:09:27+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:3.4.2-SNAPSHOT:aggregate (default-cli) on project mjavadoc494: An error has occurred in Javadoc report generation: org.apache.maven.plugins.javadoc.its:lib:jar:1.0 was not found in http://localhost:51146 during a previous attempt. This failure was cached in the local repository and resolution is not reattempted until the update interval of mrm-maven-plugin has elapsed or updates are forced -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:3.4.2-SNAPSHOT:aggregate (default-cli) on project mjavadoc494: An error has occurred in Javadoc report generation: org.apache.maven.plugins.javadoc.its:lib:jar:1.0 was not found in http://localhost:51146 during a previous attempt. This failure was cached in the local repository and resolution is not reattempted until the update interval of mrm-maven-plugin has elapsed or updates are forced...
```

**Comment by kwin** [comment]
Last IT fixed, squashed and rebased. Ready for final review.

**Review by michael-o — APPROVED** [review]

**Review by cstamas — APPROVED** [review]

File: pom.xml [hunk 4]
```diff
@@ -171,38 +171,42 @@ under the License.
...
       <artifactId>maven-plugin-annotations</artifactId>
       <scope>provided</scope>
     </dependency>
     <dependency>
-      <groupId>org.codehaus.plexus</groupId>
-      <artifactId>plexus-component-annotations</artifactId>
-      <scope>provided</scope>
+      <groupId>javax.inject</groupId>
+      <artifactId>javax.inject</artifactId>
+      <version>1</version>
```
[the remark this answers is not in the document]
Comment (hazendaz) [reply pom.xml:208]: @kwin It did get other versions, the GAV changed...

Latest in 1.x is

```
<dependency>
  <groupId>jakarta.inject</groupId>
  <artifactId>jakarta.inject-api</artifactId>
  <version>1.0.5</version>
</dependency>
```

The original has problems with fact it has no meta data.  I forget specific issue that caused in the past.  Would be better to use the officially supported one though regardless.  Note this is still javax internals.

[the remark this answers is not in the document]
Comment (kwin) [reply pom.xml:208]: I just depend on the artifact which is embedded (https://github.com/apache/maven/blob/84538c9988a25aec085021c365c560670ad80f63/pom.xml#L263-L267) and exported (https://github.com/apache/maven/blob/84538c9988a25aec085021c365c560670ad80f63/maven-core/src/main/resources/META-INF/maven/extension.xml#L96) by Maven Core. If there is an issue with that dependency we should fix this upstream. I didn't face any particular issue due to this, what metadata is missing?
Please report a bug at https://issues.apache.org/jira/projects/MNG/issues if there is an issue with this dependency.

[the remark this answers is not in the document]
Comment (hazendaz) [reply pom.xml:208]: ok thanks, the original jar has no meta-inf.  If I recall, tools that scan jars such as tomcat have issues with that jar.  Its better to use the correct api though that jakarta released for javax stream.  Makes sense to fix up stream and anywhere else and exclude the other if any legacy brings it in.

**Comment by michael-o** [comment]
I knew, I had premonition that this PR does too much. Replacing MAT has now caused: https://issues.apache.org/jira/browse/MJAVADOC-742

**Comment by jira-importer** [comment]
Resolve #1104

**Pull request merged**